### PR TITLE
Set weight ref to support share params at tensor level

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?

Set weight ref to support share params at tensor level

## How was this patch tested?
unit test

### CODE REVIEW

**Comment by wzhongyuan** [comment]
@yangw1234   Please review the change, thanks

**Comment by jason-dai** [comment]
I think it is somewhat dangerous to change val to var.

**Comment by wzhongyuan** [comment]
@yangw1234  close it for now

**Comment by yiheng** [comment]
@jason-dai maybe we can set it to a private var and the modification method can only be accessed in a limited scope. I think there's potential issues if we share weight only share storage

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -354,6 +354,9 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
    */
   def parameters(): (Array[Tensor[T]], Array[Tensor[T]]) = null
 
+
+  def setParameters(params : Array[Tensor[T]]): Unit = {}
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: private[bigdl]?

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: should we group the parameter relate API to a trait? As the AbstractModule is too large now...

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Add.scala [hunk 2]
```diff
@@ -32,7 +32,7 @@ import scala.reflect.ClassTag
 class Add[T: ClassTag](val inputSize: Int
   )(implicit ev: TensorNumeric[T]) extends TensorModule[T] with Initializable {
 
-  val bias = Tensor[T](inputSize)
+  var bias = Tensor[T](inputSize)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Add.scala]: -> private var _bias
def bias = _bias

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Add.scala]: same as the below

**Comment by yangw1234** [comment]
[https://github.com/torch/nn/blob/master/doc/module.md#sharemlps1s2sn](https://github.com/torch/nn/blob/master/doc/module.md#sharemlps1s2sn)
Can we do something like this?

1. Make parameter private
2. Do **not** provide set method, so that the parameter can only be modified by `clone`/`share`

**Pull request closed without merging**